plugins/TauMVAProducer.cc

```diff
@@ -0,0 +1,371 @@
+    iEvent.getByToken(pfCandToken_, pfcands_);
+    iEvent.getByToken(jetToken_, jets_);
+
+    int p_piplus = 211;
+
+    std::vector<float> pt, eta, phi, mass, dz, fromPV, chiso0p1, chiso0p2, chiso0p3, chiso0p4, totiso0p1, totiso0p2, totiso0p3, totiso0p4, trackiso, nearphopt, nearphoeta, nearphophi, nearestTrkDR, contJetDR, contJetCSV;
+    std::vector<int> contJetIndex;
+
+    auto selCandPf = std::make_unique<PtrVector<reco::Candidate>>();
+
+    std::cout << "Starting to process the PFCandidate Variables" << std::endl;
+
+    for (unsigned int ic = 0; ic < pfcands_->size(); ic++) {
+	const pat::PackedCandidate* pfc = &pfcands_->at(ic);
+	if (pfc->pt() < candptMin_ || fabs(pfc->eta()) > candetaMax_) continue;
+
+	if(!isA(p_piplus, pfc->pdgId())) continue; // only save charged hadrons unless otherwise specified
+
+	float chiso0p1_ = computePFIsolation(pfc, 0.0, 0.1, 0);
+	float chiso0p2_ = computePFIsolation(pfc, 0.0, 0.2, 0);
+	float chiso0p3_ = computePFIsolation(pfc, 0.0, 0.3, 0);
+	float chiso0p4_ = computePFIsolation(pfc, 0.0, 0.4, 0);
+	float totiso0p1_ = computePFIsolation(pfc, 0.0, 0.1, 1);
+	float totiso0p2_ = computePFIsolation(pfc, 0.0, 0.2, 1);
+	float totiso0p3_ = computePFIsolation(pfc, 0.0, 0.3, 1);
+	float totiso0p4_ = computePFIsolation(pfc, 0.0, 0.4, 1);
+	float trackiso_ = TrackIso(pfc,0.3,0.1);
+	int photonIndex_ = getNearPhotonIndex(pfc);
+	float nearesttrkdr_ = getDRNearestTrack(pfc);
+	
+	int jetIndex_ = getContainingJetIndex(pfc);
+	const pat::Jet* jet = jetIndex_ > -1 ? &jets_->at(jetIndex_) : 0;
+	bool jetmatch = jetIndex_ > -1 && jet->pt() > 30.0 && fabs(jet->eta()) < 2.4;
+	
+	float contjetdr_ = jetmatch ? deltaR(pfc->p4(), jet->p4()) : -1.0;
+	float contjetcsv_ = jetmatch ? jet->bDiscriminator("pfCombinedInclusiveSecondaryVertexV2BJetTags") : -1.0;
+
+	pt.push_back(pfc->pt());
+	eta.push_back(pfc->eta());
+	phi.push_back(pfc->phi());
+	mass.push_back(pfc->mass());
+	dz.push_back(pfc->dz());
+	fromPV.push_back(pfc->fromPV());
+	chiso0p1.push_back(chiso0p1_);
+	chiso0p2.push_back(chiso0p2_);
+	chiso0p3.push_back(chiso0p3_);
+	chiso0p4.push_back(chiso0p4_);
+	totiso0p1.push_back(totiso0p1_);
+	totiso0p2.push_back(totiso0p2_);
+	totiso0p3.push_back(totiso0p3_);
+	totiso0p4.push_back(totiso0p4_);
+	trackiso.push_back(trackiso_);
+	nearphopt.push_back(photonIndex_ > -1 ? pfcands_->at(photonIndex_).pt() : -1.0);
+	nearphoeta.push_back(photonIndex_ > -1 ? pfcands_->at(photonIndex_).eta() : -1.0);
+	nearphophi.push_back(photonIndex_ > -1 ? pfcands_->at(photonIndex_).phi() : -1.0);
+	nearestTrkDR.push_back(nearesttrkdr_);
+	contJetIndex.push_back(jetIndex_);
+
+	//Store in the tree jet pt -> 20 so that in the next round we can go that low
+	jetmatch = jetIndex_ > -1 && jet->pt() >= 20.0 && fabs(jet->eta()) < 2.4;
+	contjetdr_ = jetmatch ? deltaR(pfc->p4(), jet->p4()) : -1.0;
+	contjetcsv_ = jetmatch ? jet->bDiscriminator("pfCombinedInclusiveSecondaryVertexV2BJetTags") : -1.0;
```

Comment:
ContjetDr and contjetCSV: since you will store the matched jet index, these information can be recalculate on the fly. I don't think you need to store them

Reply:
I agree. I'll comment them.